contracts/misc/P2PPairStaking.sol

```diff
@@ -0,0 +1,618 @@
+                    orderHash,
+                    listingOrder.v,
+                    listingOrder.r,
+                    listingOrder.s
+                ),
+                "invalid signature"
+            );
+        }
+    }
+
+    function _validateApeOrder(ListingOrder calldata apeOrder) internal view {
+        _validateOrderBasicInfo(apeOrder);
+
+        address expectedToken = bayc;
+        if (
+            apeOrder.stakingType == StakingType.MAYCStaking ||
+            apeOrder.stakingType == StakingType.MAYCPairStaking
+        ) {
+            expectedToken = mayc;
+        }
+        require(apeOrder.token == expectedToken, "ape order invalid token");
+        require(
+            IERC721(expectedToken).ownerOf(apeOrder.tokenId) ==
+                apeOrder.offerer,
+            "ape order invalid owner"
+        );
+    }
+
+    function _validateBakcOrder(ListingOrder calldata bakcOrder) internal view {
+        _validateOrderBasicInfo(bakcOrder);
+
+        require(bakcOrder.token == bakc, "bakc order invalid token");
+        require(
+            IERC721(bakc).ownerOf(bakcOrder.tokenId) == bakcOrder.offerer,
+            "bakc order invalid owner"
+        );
+    }
+
+    function _validateApeCoinOrder(ListingOrder calldata apeCoinOrder)
+        internal
+        view
+    {
+        _validateOrderBasicInfo(apeCoinOrder);
+        require(apeCoinOrder.token == apeCoin, "ape coin order invalid token");
+    }
+
+    function _validateOrderSignature(
+        address signer,
+        bytes32 orderHash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) internal view returns (bool) {
+        return
+            SignatureChecker.verify(
+                orderHash,
+                signer,
+                v,
+                r,
+                s,
+                DOMAIN_SEPARATOR
+            );
+    }
+
+    function onERC721Received(
+        address,
+        address,
+        uint256,
+        bytes memory
+    ) external pure returns (bytes4) {
+        return this.onERC721Received.selector;
+    }
+
+    function setMatchingOperator(address _matchingOperator) external onlyOwner {
+        require(_matchingOperator != address(0), "zero address");
+        address oldOperator = matchingOperator;
+        if (oldOperator != _matchingOperator) {
+            matchingOperator = _matchingOperator;
+            emit MatchingOperatorUpdated(oldOperator, _matchingOperator);
+        }
+    }
+
+    function setCompoundFee(uint256 _compoundFee) external onlyOwner {
+        require(
+            _compoundFee < PercentageMath.HALF_PERCENTAGE_FACTOR,
+            "Fee Too High"
+        );
+        uint256 oldValue = compoundFee;
+        if (oldValue != _compoundFee) {
+            compoundFee = _compoundFee;
+            emit CompoundFeeUpdated(oldValue, _compoundFee);
+        }
+    }
+
+    function claimCompoundFee(address receiver) external onlyOwner {
+        this.claimCApeReward(receiver);
```

Comment:
when we claim the compound fee, only owner can call it. but the function `claimCApeReward` claims the reward for `msg.sender` which will be the `owner()` and not `address(this)`we should probably make an internal function called _ claimCApeReward and send the the address of the claimer.